ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala

```diff
@@ -36,40 +37,50 @@ object ContractDao {
   def initialize(implicit log: LogHandler): ConnectionIO[Unit] =
     Queries.dropAllTablesIfExist *> Queries.initDatabase
 
-  def lastOffset(party: domain.Party, templateId: domain.TemplateId.RequiredPkg)(
-      implicit log: LogHandler): ConnectionIO[Option[domain.Offset]] =
+  def lastOffset(parties: OneAnd[Set, domain.Party], templateId: domain.TemplateId.RequiredPkg)(
+      implicit log: LogHandler): ConnectionIO[Map[domain.Party, domain.Offset]] = {
+    import doobie.postgres.implicits._
     for {
       tpId <- Queries.surrogateTemplateId(
         templateId.packageId,
         templateId.moduleName,
         templateId.entityName)
-      offset <- Queries.lastOffset(party.unwrap, tpId).map(_.map(domain.Offset(_)))
+      offset <- Queries
+        .lastOffset(domain.Party.unsubst(parties), tpId)
+        .map(_.map { case (k, v) => (domain.Party(k), domain.Offset(v)) })
     } yield offset
+  }
 
   def updateOffset(
-      party: domain.Party,
+      parties: OneAnd[Set, domain.Party],
       templateId: domain.TemplateId.RequiredPkg,
       newOffset: domain.Offset,
-      lastOffset: Option[domain.Offset])(implicit log: LogHandler): ConnectionIO[Unit] =
+      lastOffsets: Map[domain.Party, domain.Offset])(
+      implicit log: LogHandler): ConnectionIO[Unit] = {
+    import cats.implicits._
+    import doobie.postgres.implicits._
+    import scalaz.OneAnd._
+    import scalaz.std.set._
+    import scalaz.syntax.foldable._
+    val partyVector = domain.Party.unsubst(parties.toVector)
     for {
       tpId <- Queries.surrogateTemplateId(
         templateId.packageId,
         templateId.moduleName,
         templateId.entityName)
-      rowCount <- Queries.updateOffset(
-        party.unwrap,
-        tpId,
-        newOffset.unwrap,
-        lastOffset.map(_.unwrap))
-      _ <- if (rowCount == 1)
+      rowCount <- Queries.updateOffset(partyVector, tpId, newOffset.unwrap, lastOffsets.map({
```

Comment:
@S11001001 is there a two-parameter version of `unsubst`? I couldn’t figure out how to strip the tag on both parameters.

Reply:
Er, I missed the context and responded up there https://github.com/digital-asset/daml/pull/7571/files#r501203465 , but the technique is the same. A conveniently inferrable subst or unsubst can be derived for any specific shape of type function, including "first param of a 2-param type constructor".